### DISCUSSION TITLE
Turbo Button line

### DISCUSSION THREAD

**Discussion by BrandoJotaro** [ndeadly/MissionControl#425] [Q&A]
What would a line to make a button repeat itself look like? I'm trying to add one so when I hit (A) ones, it will be pressed many times in a short time.

**Comment by ndeadly** [comment]
The simplest way would be to add some code in [this function](https://github.com/ndeadly/MissionControl/blob/0ee33b355b2d8649dddcb754ca928d15778636de/mc_mitm/source/controllers/emulated_switch_controller.cpp#L200-L218) to toggle the state of the button on/off every time an input report is received. This would require also storing the previous button state.

Something like

```
if (m_buttons.A) {
    switch_report->input0x30.buttons.A = !m_buttons_prev.A
}

m_buttons_prev = m_buttons;
```

@Banz99 has done some experiments with this, but I'm not sure he's made anything public. It works with the controllers he has, however I'm not totally convinced it's the right approach given that the rate of button state switching is highly dependent on the rate that the controller is sending updates. This would mean that some controllers may be able to turbo faster than others, and these differences could also be affected by the rate at which a given games polls for inputs. I'm also pretty sure that there are controllers out there that only send input reports when the the state of the inputs actually changes, rather than a constant stream. Such controllers could not work at all with the above approach. These inconsistencies are the main reason I haven't considered such a feature for inclusion in a release so far.

**Reply by BrandoJotaro** [reply]
Thank you. I tried using this line but I always get errors when compiling that the "buttons" part of "buttons->ZL" was not declared in this scope.

**Reply by ndeadly** [reply]
No idea without seeing the code that you added. There is nothing called `buttons` in my code though. So unless you added this variable yourself, the compiler error is correct - it hasn't been declared. It's either `switch_report->input0x30.buttons` or `m_buttons`, and the actual button value would be accessed with `.` not `->`

**Reply by BrandoJotaro** [reply]
Even when using m_buttons it still doesn't work :(

**C:/Users/Jotaro/Desktop/MissionControl/MissionControl/mc_mitm/source/controllers/switch_controller.cpp:95:13: error: 'm_buttons' was not declared in this scope
   95 |         if (m_buttons.ZL) {
      |             ^~~~~~~~~
C:/Users/Jotaro/Desktop/MissionControl/MissionControl/mc_mitm/source/controllers/switch_controller.cpp:96:52: error: 'm_buttons_prev' was not declared in this scope
   96 |             switch_report->input0x30.buttons.ZL = !m_buttons_prev.ZL
      |                                                    ^~~~~~~~~~~~~~
C:/Users/Jotaro/Desktop/MissionControl/MissionControl/mc_mitm/source/controllers/switch_controller.cpp:99:9: error: 'm_buttons_prev' was not declared in this scope
   99 |         m_buttons_prev = m_buttons;
      |         ^~~~~~~~~~~~~~
C:/Users/Jotaro/Desktop/MissionControl/MissionControl/mc_mitm/source/controllers/switch_controller.cpp:99:26: error: 'm_buttons' was not declared in this scope
   99 |         m_buttons_prev = m_buttons;**

This is the line I added for pro controller in order to have ZL as a turbo button:

**if (m_buttons.ZL) {
            switch_report->input0x30.buttons.ZL = !m_buttons_prev.ZL
        }

        m_buttons_prev = m_buttons;**

Could you please help me

**Reply by ndeadly** [reply]
`m_buttons` doesn't exist in the `SwitchController` class found in `switch_controller.hpp`, it is only found in `EmulatedSwitchController`. This is because `SwitchController` represents an official controller and doesn't need to store the converted button format because it's already in the correct form. If you want this to work for official controllers too you will need to move `m_buttons` from `EmulatedSwitchController` to the parent `SwitchController` class. You also need to declare `m_buttons_prev` yourself. It doesn't exist in my code currently.
